### ISSUE TITLE
re-assess the design decision to have all single-label DNS requests effectively get black-holed

### ISSUE THREAD

**Issue by fermulator** [systemd/systemd#13763]
> systemd --version
systemd 237
+PAM +AUDIT +SELINUX +IMA +APPARMOR +SMACK +SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +SECCOMP +BLKID +ELFUTILS +KMOD -IDN2 +IDN -PCRE2 default-hierarchy=hybrid

**Used distribution**
 > Ubuntu 18.04.3 LTS

**Expected behaviour you didn't see**
 > As per https://github.com/systemd/systemd/issues/10298, local single-label network name resolution MUST continue to work.  That is, in networks that have DNS pointing to a local server on the same subnet, we SHOULD continue to forward requests there, but NOT if it is external/public.

> Further, although perhaps confounded, in https://github.com/systemd/systemd/issues/2514, clearly some decisions were made for technical and purist reasons (or misinterpretation of the standard) - resulting in requirements/functionality breaking

**Unexpected behaviour you saw**
 > systemd-resolv is unable to resolve single label nameservers via the local DNS, but if we ask the local DNS directly via `nslookup` it works

Consider, the default for systemd-resolv, I will use "fermmy-git" as my example server, which SHOULD resolve to 10.0.0.54.
```
 nslookup fermmy-git
Server:		127.0.0.53
Address:	127.0.0.53#53

** server can't find fermmy-git: SERVFAIL
```

Compare with, how it SHOULD resolve if we ask the router directly:
```
$ nslookup fermmy-git 10.0.0.1
Server:		10.0.0.1
Address:	10.0.0.1#53

Name:	fermmy-git
Address: 10.0.0.54
```

**Steps to reproduce the problem**

Start with the very common Ubuntu distribution, <=16.04, that leverages resolvconf buy default.
Your network shall have a router, with static or DHCP where all systems point their primary DNS to the router.
The router has some internal nameservers (e.g. static DHCP entries), for systems within the network to do name-IP-resolution.
e.g. "my-server" should resolve to "192.168.0.10" (or whatever)

Upgrade to 18.04 (start using `systemd-resolv`).

If we use all the standard tools (`ping`, `nslookup`, `systemd-resolve`, ...) -but _especially_ from web browsers or applications that are hooked in by local name ... they will not resolve.  We have broken backwards compatibility for this very standard network design.

----
Other refs (as I find them)
* https://askubuntu.com/questions/917784/systemd-resolved-does-not-query-dns-server-for-local-domain
* https://github.com/systemd/systemd/issues/6224

**Comment by fermulator** [comment]
https://github.com/systemd/systemd/issues/6224 complains about precisely the same thing.

**Comment by fermulator** [comment]
What is the discussion protocol around such a situation as this? Is there a frequent meeting where stakeholders gather at a specific time to discuss in an agenda?

**Comment by thiagocmartinsc** [comment]
I'm facing many, many problems with name resolution in my networks based on Ubuntu 18.04. I really like some ideas of `systemd` but, this `systemd-resolved` thing is a complete disaster.

For example, out of blue, I can't `ping my-server.mydomain.com` anymore, forcing me to re-run `netplan apply` to be able to resolve my own domains again, then, it stops working again, another `netplan apply`.

On this very same Network, systems without systemd works just fine, like Devuan, MacOS, Windows, CentOS 6, old Ubuntu with `upstart`, ONLY systemd-based machines are failing badly.

I also do not wish to leak queries public DNSes, since my DHCP sets my own internal DNS Servers as `nameservers` for my Ubuntu Desktops, Servers, Macbooks and etc.

What, is, happening here?!    :-(

A google about "systemd resolved issues" have a million results... Come on @poettering, why are you doing this?!

I'm still trying to make `systemd-resolved` to work (trying for days already), the last things that I found was:

1- install `libnss-resolve` (didn't tested yet);
2- Point `/etc/resolv.conf` to `/run/systemd/resolve/resolv.conf`, instead of default `../run/systemd/resolve/stub-resolv.conf`


BTW, I'm sorry for posting this here, thing is that I honestly don't know what to do anymore, neither which Github issue to interact (there are so many resolved issues out there)...

It's kinda hard to get rid of `systemd-resolved` on Ubuntu because it defaults to Netplan, which depends on `networkd`/`reesolved`. I don't have Network Manager.

**Comment by eddyg** [comment]
It *seems* like using the [`Domains=~.` construct in `resolved.conf`](https://github.com/systemd/systemd/commit/2df225294#diff-7fafbb4ebb0047d70a37c604a941252fR483-R487) is meant to address this issue.

However, having this configured doesn't seem to make a difference for me, so I'm not sure the special `~.` construct is doing what I _want_ it to do.

> Note: I'm running Ubuntu 18.04, so it has `systemd-resolve` (*systemd 237*) and not `resolvectl` (*systemd 239*):

```
root@laptop:# systemd-resolve --status
Global
         DNS Servers: 10.0.0.1
          DNS Domain: ~.
          ...

root@laptop:# systemd-resolve docker.example.com
docker.example.com: 10.0.0.36

-- Information acquired via protocol DNS in 156.0ms.
-- Data is authenticated: no

root@laptop:# systemd-resolve docker
docker: resolve call failed: No appropriate name servers or networks for name found
```

(Note that I *could* set `example.com` as a search domain, but that's *not* the solution I'm looking for... I'd like *unqualified* DNS names to be sent to the configured name server, like they always have in the past.)

Any suggestions on when `Domains=~.` *would* have an effect?

**Comment by vontrapp** [comment]
I've said this in whatever other issue thread and I'll say it again. The standard cited for this decision does NOT mean to preclude the sending of single label queries to dns for resolution.

The case is specifically covering the forwarding of queries from the DNS server facing the client which received the query. A DNS *server* shall not forward a single label query in a recursive lookup. This is intended to not hit the *root* servers with single label lookups.

It was NOT meant for the case if a client querying that clients local configured DNS server!!

Please change this back!

**Comment by jcartner-young** [comment]
I will now speak the immortal words: "_I am also having this issue_."

I should rather say 'HAD' this issue - I, just this minute, disabled the awful excuse for a DNS resolver, systemd-resolved, and installed good ol' unbound and reconfigured my resolv.conf and network-manager configuration.

Everything now resolves perfectly fine without errors, and (not sure if anyone else has noticed similar responsiveness improvements) much, MUCH quicker, too.

I should've recorded some timings as evidence, but you're just going to have to take my word for it - I observed a near-zero delay in the new requests (short and FQDN queries) returning proper answers, compared to the more appreciable delay when using systemd-resolved.

I have to agree with every other person in this (and the myriad other threads about this hideous breakage in core system functionality)

BTW, I use Debian Mint v19.

**The solution I found** was simply to disable systemd-resolved, install `unbound` and reconfigure `resolv.conf` and `network-manager` : [Fix Linux DNS Issues Caused by Systemd-Resolved](https://blobfolio.com/2017/05/fix-linux-dns-issues-caused-by-systemd-resolved/)
_Just so everyone is aware of how long this issue has been going around unresolved (pun intended lol)... This blog post was written in 2017 and the author mentions it's been driving him mad for the last few years!_

**Opinions below:**

I don't want to antagonise the systemd developers [as much as systemd-resolved has antagonised me] or seem ungrateful - Coding to a high standard is hard and takes a lot of time and effort, but when the result after hours of work is broken code, you have to drop it - You can't pass it off as 'working' when it clearly isn't.

In my uninformed opinion, DNS queries should be attempted first (when an DNS IP is configured, recursing FQDNs but not short names) then failing over to LLMNR (using the DNS search domain suffix(es)) - maybe swapping DNS and LLMNR priorities on a per-domain basis?  Maybe integrating LLMNR into DNS resolution as a last-resort at a domain-local resolution? This would expand the use cases and deployment scenarios for this new DNS resolution methodology.

LLMNR strikes me as a very trusting protocol - If I have a log server and someone else starts answering LLMNR queries for _syslogserver1_ (for example) what would happen to my syslog data if my log server's name is resolved using systemd-resolverd with LLMNR? (Akin to ARP cache poisoning.)

FWIW, I'm not the biggest fan of systemd.

PS: It does seem to me that a majority of issues I've been having with various Linux's over the last few years may well be attributable to this systemd mess.  Moreover, during all this time I thought it was my naivety or inexperience of the intricacies of Linux systems... Nope.

PPS: Page load times definitely seem faster.  Could be perception bias.

**Comment by fermulator** [comment]
sooo long; I guess the workaround is to remove systemd-resolv?  No one has been able to reply with a PROPER fix or WORKAROUND.

Time to write an ansible script that removes it on all my systems. So tired of:
```
$ nslookup local-hostname <ip-of-router>
$ ssh <ip>
```

**Comment by garfunkel** [comment]
Spent a solid 12 hours on a fresh Arch install with this bloody thing trying to get it to resolve the NAS on my local network. Looking at the standard, I agree - I can't see why this decision was made in the first place, the standard never mentions blackholing local hostname resolution at all. In fact, it's awfully clear.

I feel desperately sorry for companies deploying recent Ubuntu releases on their networks making the very rather reasonable assumption that staff will be able to access their network servers like they have for the last 40 years... because they will be in for a nasty surprise.

I wasn't able to produce a working solution to this after the 12 hours of hair pulling and decided it would be better for my health prospects to just start afresh with NetworkManager.

This really needs to be fixed. It's sheer bloody-mindedness to continue to ignore this.

**Comment by fermulator** [comment]
@keszybz , I don't think updating the `man` page here is going to sufficiently resolve the problem.

There are clearly two options:
1. restore single-label DNS functionality for LOCAL queries to LOCAL DNS by default
or
2. provide clearly documented instructions for how to configure `systemd-resolve` for allowing it

**Comment by AlD** [comment]
@poettering since you added the "needs-discussion" label, do you still think this is the case?
If yes I'd be happy to contribute to providing evidence that the implementation interprets the standard incorrectly, whichever standard this decision it may be that this decision was based on.

**Comment by 3ronco** [comment]
Can't reproduce the behavior described by @fermulator.
As far as i can tell _systemd-resolved_ works exactly as expected. As soon as a proper configured DNS server returns an entry for a single-label name it gets returned by _resolved_ as well as the name server itself depending on the protocol the name was requested through and regardless of the request originating from the router or any other host on the subnet.
I'm running a multi-homed configuration with a lot of interfaces with multiple DNS servers each configured on the link it should resolve on with multiple links to different providers (configured by _DHCP_ or even _pppoe_) plus some distinct VPN connections. In combination with _systemd-networkd_ i can even switch providers dynamically by bringing down or up the specific interface and entries get resolved immediately recognizing the change of default route _but_ only when they are configured by _systemd-networkd_ which updates the entries of available DNS resolvers on the specific link.

Here's my config, on the local host in `/etc/nsswitch.conf` i'm pointing all sorts of DNS request (DNS, LLMNR, mDNS) to _resolved_.
```
[...]
hosts:     resolve
[...]
```
In `/etc/resolv.conf` it points to _resolved_ too:
```
# This file is managed by man:systemd-resolved(8). Do not edit.
#
[...]

nameserver 127.0.0.53
options edns0
search private.domain.net
```
Notice that the `search` statement still needs to append _your_ domain to requests so that _resolved_ knows where to forward that DNS request. Assumed that one of your DNSs is able to resolve that and the link has a proper Domain entry set in it's _systemd.network_ config file (!)
```
[Match]
Name=myNIC

[Network]
Address=...
DNS=x.x.x.x y.y.y.y zzzz:zzzz::z

# resolve my private domain on this link
Domains=private.domain.net

LLMNR=true
MulticastDNS=true
```
Now hosts in your subnet usally won't talk to router's localhost but to the IP of your DNS. In a common config you would usally setup some kind of forwarder eg. your provider's DNS _but_ when you want _resolved_ to select an approriate DNS then your DNS instance (eg. _Bind9_) should point back to _resolved_.
```
options {
[...]
	forwarders {
		127.0.0.53;                        // caching fallback nameserver on localhost (resolved)
	}
[...]
}
```
Of course, this kind of configuration is only possbible when your DNS is running on the host _resolved_ is running on but a scenrio with a caching forward-only DNS in between is also possible.

For router host the DNS request chain is:
`resolved -> local or DNS on specific link -> if unsuccesful -> resolved`

For other hosts on the subnet it's:
`local DNS (if unsuccesful) -> resolved -> DNS on specific link`

This setup is really bullet proof even to the point when almost all your DNSs go down with LLMNR or mDNS requests still working.

**Comment by AlD** [comment]
@3ronco see https://github.com/systemd/systemd/issues/13763#issuecomment-554754955 regarding using a search domain. It's a workaround, nothing else.

**Comment by 3ronco** [comment]
Ok, for a workaround it's purpose is exactly described in the man page in PROTOCOLS AND ROUTING: _"... Effectively this means: in order to preferably route all DNS queries not explicitly matched by search/route-only domain configuration to a specific link, configure a "~."  route-only domain on it."_
The _global_ section isn't a specific link therefore the exclusion intended may fail.

**Comment by AlD** [comment]
That part of the man page is actually outdated and has been updated by the commit above your comment (https://github.com/systemd/systemd/pull/15501/commits/ae7932ffe0459b8d4795a846ddcdc0f7f46856fd).

**Comment by fermulator** [comment]
https://github.com/systemd/systemd/issues/2514#issuecomment-562963754 has a very nice summary from @h1z1 explaining why local hostname resolution has incorrectly been broken.

**Comment by keszybz** [comment]
Resolved in 2bd5e1b272.

**Comment by fermulator** [comment]
umm... so we didn't REALLY do what this ticket was asking for; that is, to re-assess these decisions as being incorrectly interpreted from the specification.  Definitely appreciate the man updates to explain how to re-enable the functionality (allows users to workaround the problem and restore previous functionality), but I would still argue that it should be the DEFAULT!

-- re-open?

**Comment by garfunkel** [comment]
I agree, this should be re-opened.

The default behaviour of the solution should conform to the standard behaviour of what users would expect of the solution. It appears we all now agree that systemd was not correctly interpreting the standard previously, and now is - so, it stands to reason that the new/correct behaviour should be enabled by default.

**Comment by keszybz** [comment]
Sorry, the standard is pretty clear and there certainly is no agreement that it was not interpreted correctly or that it should be ignored. I implemented exactly what @fermulator asked for in https://github.com/systemd/systemd/issues/13763#issuecomment-615222664. The behaviour that was requested for can be enabled with a single switch, but it will not be made the default since it's against the standard and is something that is only appropriate in limited circumstances.

**Comment by fermulator** [comment]
sigh;  fwiw; I have abandoned `systemd-resolved` due to this disagreement

**Comment by xuto2** [comment]
why is this closed without any action?

**Comment by humanoid2050** [comment]
For the people who follow, here is what I did to get single-label lookup working (i.e. by hostname of target computer). Starting from a plain installation of a modern distribution like Ubuntu 20.04:

```
sudo apt install libnss-resolve
sudo sed -i 's/^#LLMNR=.*$/LLMNR=yes/' /etc/systemd/resolved.conf

```
Do this on every computer in your lan, or at least the ones you want to be able to find by hostname. 

The first line installs a libnss plugin to tell libnss to actually use systemd-resolved when doing lookups. Without this, programs which use glibc NSS as their lookup entrypoint will appear to not work, even if resolvectl can find the queried name. 

The second line tells systemd-resolved to fully participate in LLMNR lookups, both for searching and being found. For completeness of discussion, you could also set 'LLMNR=resolve' and install a stand alone service like llmnrd. This will cause systemd-resolved to do llmnr queries, and the llmnrd service (which is running on the remote target machine) will provide the response. IMHO, just using the systemd-resolved solution is easier as there are fewer dependencies.

Important detail: systemd-resolved will only do these LLMNR lookups if both the global and link-specific settings are not 'no'. You can query these with `resolvectl status`. The sed line above sets the global parameter. By default, the link-specific parameters are 'yes'. If this changes in the future or your distribution is different, you will need to address this. Fortunately, the default seems to be 'yes'.

Reboot your computers and enjoy.

**Comment by moodboom** [comment]
> Sorry, the standard is pretty clear and there certainly is no agreement that it was not interpreted correctly or that it should be ignored. I implemented exactly what @fermulator asked for in [#13763 (comment)](https://github.com/systemd/systemd/issues/13763#issuecomment-615222664). The behaviour that was requested for can be enabled with a single switch, but it will not be made the default since it's against the standard and is something that is only appropriate in limited circumstances.

By now, you must have some concept of how many man-years of time systemd-resolved has wasted across the globe.  Can you check your ego for one second and consider that?  Default behavior really matters.  At the same time, thanks for your work.
